plugins/calibre.koplugin/metadata.lua

```diff
@@ -0,0 +1,250 @@
+--[[
+    This module implements functions for loading, saving and editing calibre metadata files.
+
+    Calibre uses JSON to store metadata on device after each wired transfer.
+    In wireless transfers calibre sends the same metadata to the client, which is in charge
+    of storing it.
+--]]
+
+local rapidjson = require("rapidjson")
+local logger = require("logger")
+local util = require("util")
+
+local unused_metadata = {
```

Comment:
On the other hand, `user_metadata` is hands down going to be the *largest* (and most complex) structure in there, so that's potentially a strike against it... ;).

Reply:
Fair warning. I think I will leave it for somebody that uses the feature :smiley:

Reply:
Yeah, to give you an example....

```json
"user_metadata": {
      "#hyphenated": {
        "is_category": false,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 3,
        "column": "value",
        "rec_index": 22,
        "search_terms": [
          "#hyphenated"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "bool",
        "#value#": null,
        "category_sort": "value",
        "table": "custom_column_3",
        "is_editable": true,
        "label": "hyphenated",
        "display": {},
        "name": "Hyphenated"
      },
      "#koboreadpct": {
        "is_category": false,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 5,
        "column": "value",
        "rec_index": 26,
        "search_terms": [
          "#koboreadpct"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "int",
        "#value#": 100,
        "category_sort": "value",
        "table": "custom_column_5",
        "is_editable": true,
        "label": "koboreadpct",
        "display": {
          "description": "Current Reading status on Kobo",
          "number_format": null
        },
        "name": "Progress"
      },
      "#read": {
        "is_category": false,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 7,
        "column": "value",
        "rec_index": 28,
        "search_terms": [
          "#read"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "composite",
        "#value#": "Yes",
        "category_sort": "value",
        "table": "custom_column_7",
        "is_editable": true,
        "label": "read",
        "display": {
          "description": "Read status",
          "composite_template": "program:\n#\tIf it's Read (i.e. at 100%) on Kobo, flag it as read, otherwise, use the Kindle status as-is.\n\ttest(field('#koboreadpct'), cmp(field('#koboreadpct'), 100, '', 'Yes', ''), field('#kindleread'));",
          "composite_sort": "bool",
          "contains_html": false,
          "make_category": false,
          "use_decorations": 2                                                                                                                                                                                                                                                                                                                                           
        },
        "name": "Read"
      },
      "#kindleread": {
        "is_category": false,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 1,
        "column": "value",
        "rec_index": 23,
        "search_terms": [
          "#kindleread"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "bool",
        "#value#": null,
        "category_sort": "value",
        "table": "custom_column_1",
        "is_editable": true,
        "label": "kindleread",
        "display": {
          "description": "Read status on Kindle"
        },
        "name": "Read on Kindle"
      },
      "#kobolastread": {
        "is_category": false,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 8,
        "column": "value",
        "rec_index": 24,
        "search_terms": [
          "#kobolastread"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "datetime",
        "#value#": "2020-06-03T06:08:34+00:00",
        "category_sort": "value",
        "table": "custom_column_8",
        "is_editable": true,
        "label": "kobolastread",
        "display": {
          "date_format": null,
          "description": "Last time the book was Read on Kobo"
        },
        "name": "Last Read"
      },
      "#koboreadloc": {
        "is_category": true,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 6,
        "column": "value",
        "rec_index": 25,
        "search_terms": [
          "#koboreadloc"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "text",
        "#value#": null,
        "category_sort": "value",
        "table": "custom_column_6",
        "is_editable": true,
        "label": "koboreadloc",
        "display": {
          "description": "Current Reading Location on Kobo",
          "use_decorations": 0                                                                                                                                                                                                                                                                                                                                           
        },
        "name": "Reading Location"
      },
      "#pages": {
        "is_category": false,
        "#extra#": null,
        "kind": "field",
        "is_custom": true,
        "is_csp": false,
        "colnum": 2,
        "column": "value",
        "rec_index": 27,
        "search_terms": [
          "#pages"
        ],
        "link_column": "value",
        "is_multiple2": {},
        "is_multiple": null,
        "datatype": "int",
        "#value#": 288,
        "category_sort": "value",
        "table": "custom_column_2",
        "is_editable": true,
        "label": "pages",
        "display": {
          "number_format": "{0:,}"
        },
        "name": "Pages"
      }
    },
```

:oO:

Reply:
Granted, ideally, one would not deal with my insane composite column, and instead only something like #kindleread (which I maintained manually), or #koboreadpct (which is automatically kept in sync w/ Nickel via the Kobo driver).

But, still, potentially neat, but probably a huge can of worms ^^.

Reply:
And that's still considered a fairly tame use of custom columns ^^.

Reply:
Wow. I'm starting to understand what a calibre user looks like.

I just use it to keep purchased ebooks (deDRMed). I had to download a bunch of Gutenberg books to test this PR since my purchases are not so many. (~ 200 books).